text/0000-inline-asm.md

```diff
@@ -0,0 +1,873 @@
+
+Let's take our previous example of an `add` instruction:
+
+```rust
+let mut a = 4;
+let b = 4;
+unsafe {
+    asm!(
+        "add {0}, {1}",
+        inlateout(reg) a, in(reg) b,
+        flags(pure, nomem, nostack)
+    );
+}
+assert_eq!(a, 8);
+```
+
+Flags can be provided as an optional final argument to the `asm!` macro. We specified three flags here:
+- `pure` means that the asm code has no observable side effects and that its output depends only on its inputs. This allows the compiler optimizer to call the inline asm fewer times or even eliminate it entirely.
+- `nomem` means that the asm code does not read or write to memory. By default the compiler will assume that inline assembly can read or write any memory address that is accessible to it (e.g. through a pointer passed as an operand, or a global).
+- `nostack` means that the asm code does not push any data onto the stack. This allows the compiler to use optimizations such as the stack red zone on x86_64 to avoid stack pointer adjustments.
+
+These allow the compiler to better optimize code using `asm!`, for example by eliminating pure `asm!` blocks whose outputs are not needed.
+
+See the reference for the full list of available flags and their effects.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Inline assembler is implemented as an unsafe macro `asm!()`.
+The first argument to this macro is a template string literal used to build the final assembly.
+The following arguments specify input and output operands.
+When required, flags are specified as the final argument.
+
+The following ABNF specifies the general syntax:
+
+```
+dir_spec := "in" / "out" / "lateout" / "inout" / "inlateout"
+reg_spec := <arch specific register class> / "<arch specific register name>"
+operand_expr := expr / "_" / expr "=>" expr / expr "=>" "_"
+reg_operand := dir_spec "(" reg_spec ")" operand_expr
+operand := reg_operand / "imm" const_expr / "sym" path
+flag := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn"
+flags := "flags(" flag *["," flag] ")"
+asm := "asm!(" format_string *("," [ident "="] operand) ["," flags] ")"
+```
+
+[format-syntax]: https://doc.rust-lang.org/std/fmt/#syntax
+
+## Template string
+
+The assembler template uses the same syntax as [format strings][format-syntax] (i.e. placeholders are specified by curly braces). The corresponding arguments are accessed in order, by index, or by name. However, implicit named arguments (introduced by [RFC #2795][rfc-2795]) are not supported.
+
+The assembly code syntax used is that of the GNU assembler (GAS). The only exception is on x86 where the Intel syntax is used instead of GCC's AT&T syntax.
```

Comment:
Does this imply that `x86` will use the format of `mnemonic	destination, source` (like Intel Syntax) while every other platform will use `mnemonic	source, destination` (like AT&T/GAS)? Because that'll extremely hurt portability/readability.

Reply:
Every other architecture already uses `mnemonic destination, source`. x86 is the exception with AT&T syntax that reverses this order.

Reply:
Okay, so every platform is standardized to `mnemonic dst, src`?

This is a bit confusing, as I'm not sure if the Rust-Source is being specified, or the IR which is being transferred to the compiler. 

So does this imply that on _non-x86_ platforms `rustc` will be responsible for re-ordering `mnemonic` arguments to the backend?

Reply:
There is no reordering needed, all non-x86 platforms already use the correct ordering. The Intel/ATT syntax split is only a thing on x86.